### PR TITLE
Extend NMSG Unit Tests

### PR DESCRIPTION
**Pull request by kadealicious**
LCOV coverage report shows the following increase in coverage: 
- Line coverage: 58.9% -> 67.6%
- Function coverage: 68.0% -> 84.3%

This MR contains the following changes: 
- Add LCOV coverage report option at configuration time.
- Add 5 new unit tests for message modules, socket fragmentation, ZMQ endpoints, kicker files, and daemonization.
- Remove some unused IP fragment reassembly getter functions.
- Remove auto-generated protobuf-c files from code coverage output.
- Create new (defanged) sample data files.

To generate a coverage report for `nmsg-1.2.0` (the current version on the `next` branch): 
```
make -j
make dist
tar -xvf nmsg-1.2.0.tar.gz
rm nmsg-1.2.0.tar.gz
cd nmsg-1.2.0
./configure --with-coverage
make -j
make check -j
make report-coverage -j
your-browser coverage-html/index.html
```

### CODE REVIEW

**Comment by shw700** [comment]
These unit tests (`make check`) fail for me on my laptop. I just confirmed that the same is NOT the case for either the `master` or `next` branches.

**Comment by kadealicious** [comment]
> These unit tests (`make check`) fail for me on my laptop. I just confirmed that the same is NOT the case for either the `master` or `next` branches.

This has been addressed.

File: tests/test-daemon.sh [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	killprocess_pid=$!
+	wait $killprocess_pid
+}
+
+echo "Testing nmsgtool daemonization: "
+
+# Start a daemonized nmsgtool and wait for it to finish.
+$nmsgtool_test -ddddd -D -r $infile -w $outfile &
+sleep 1
+killproc $!
```
Comment (shw700) [inline tests/test-daemon.sh:34]: Note that this kill doesn't really do anything. If there were no `-D` parameter then the special value `$!` would correctly point to what you want it to, but the first step of daemonization is to `fork()`, allowing the parent process to exit while continuing to execute in the child. So here the return value of `$!` will point to the stale parent pid that has already exited, and not the new process that becomes the running nmsgtool instance that you're interested in.

File: tests/test-daemon.sh [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+$($nmsgtool_test -l 127.0.0.1/12345 -D -P $pidfile)
+sleep 1
+proc_pid=$(pgrep -x "nmsgtool" | tail -n 1)
+file_pid=$(cat $pidfile | head -n 1)
+
+# Compare actual PID with PID in file.
+[[ "$proc_pid" -eq "$file_pid" ]]
+check "PID in file ($file_pid) equals actual PID ($proc_pid)"
+
+# Start an nmsgtool and attempt to change users.  This will fail because we
```
Comment (shw700) [inline tests/test-daemon.sh:50]: I'm not sure this is a useful test since the default application behavior is NOT to enact any kind of permissions change, and we never attempt to do so successfully.

File: tests/test-kicker.sh [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+echo "Testing kicker: "
+
+mkdir $outdir
+
+# Read input data with our kicked nmsgtools.
+$nmsgtool_test -ddddd -j $infile -c 1 -k "$kicker" > $outfile
+file_count=$(wc -l $outfile | awk '{print $1}')
+entry_comparison=0; [ "$file_count" == "15" ] && entry_comparison=1
```
Comment (shw700) [inline tests/test-kicker.sh:33]: This notation is a bit cryptic... I think simply `[ "$file_count" == "15" ]` should suffice, no?

Comment (shw700) [reply tests/test-kicker.sh:33]: This is indeed using the kicker functionality under the hood but is it really testing it? The input file has 15 lines; the output has 15 lines. For all we know, it could simply be regurgitating the input without applying any transformation. I think a more reasonable test might be running this in a fresh test directory, making sure that directory is first empty, and then checking to see that 15 files have been created which match a glob of say... the current year/month/day/etc. (and of course, deleting them all afterwards)

File: tests/test-socket-frag.sh [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+# Wait for the messages to send/receive.
+sleep 1
+
+# Kill em all!
+kill $writer_pid >/dev/null 2>&1
+killwriter_pid=$!
+kill $listener_pid >/dev/null 2>&1
+killlistener_pid=$!
+wait $killwriter_pid
```
Comment (shw700) [inline tests/test-socket-frag.sh:47]: You should just be able to wait on `$writer_pid` and `$listener_pid` directly...

Comment (shw700) [reply tests/test-socket-frag.sh:47]: Also, are you sure we can't just wait on them instead of killing them?

File: tests/test-msgmod.sh [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	final_entry_comparison=0; [ "$line_count" == "$3" ] && final_entry_comparison=1
+	check "base/$1 pcap message parsing ($line_count/$3 payloads parsed)"
+}
+
+echo "Testing message modules: "
+
+mkdir $outdir
+
+# Test message modules (setting a few environment variables to increase coverage).
+export DNSQR_CAPTURE_RD=1
```
Comment (shw700) [inline tests/test-msgmod.sh:39]: Exercising functionality but not testing it will just contribute to the illusion that certain code paths are properly tested when in fact they are not.

File: tests/test-daemon.sh (hunk 1 above)
Comment (kadealicious) [reply tests/test-daemon.sh]: Fixed

File: tests/test-daemon.sh [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+killproc $!
+
+# Verify that the number of entries output by the daemon is correct.
+diff $infile $outfile
+check "diff on daemon input vs. output"
+
+# Start a daemonized nmsgtool that outputs a file containing its PID.
+$($nmsgtool_test -l 127.0.0.1/12345 -D -P $pidfile)
+sleep 1
+proc_pid=$(pgrep -x "nmsgtool" | tail -n 1)
```
[the remark this answers is not in the document]
Comment (kadealicious) [reply tests/test-daemon.sh]: Fixed

File: tests/test-daemon.sh (hunk 2 above)
Comment (kadealicious) [reply tests/test-daemon.sh]: Removed

File: tests/test-socket-frag.sh (hunk 4 above)
Comment (kadealicious) [reply tests/test-socket-frag.sh]: We can (and now do) wait on them instead of killing them.

File: tests/test-msgmod.sh (hunk 5 above)
Comment (kadealicious) [reply tests/test-msgmod.sh]: Fixed

File: tests/test-daemon.sh [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+check() {
+	if [ $? = "0" ]; then
+		echo "PASS: $*"
+	else
+		echo "FAIL: $*"
+		retval=1
+	fi
+}
+
+killproc () {
```
Comment (shw700) [inline tests/test-daemon.sh:23]: This can be removed. A simple invocation of `kill` should do the trick

Comment (kadealicious) [reply tests/test-daemon.sh]: Fixed

File: tests/test-misc.c [hunk 8]
```diff
@@ -479,71 +477,113 @@ test_pcap(void)
...
+	const char bpf_filter_string[] = "tcp dst port 80 or tcp src port 80";
 	struct timespec ts;
-	struct pcap_pkthdr *pphdr;
-	const uint8_t *pkdata;
+	struct nmsg_ipdg ni;
 
-	check_return(nmsg_pcap_input_setfilter_raw(pcap, BPF_FILTER_STRING) == nmsg_res_success);
-//	check_return(nmsg_pcap_input_setfilter(pcap, BPF_FILTER_STRING) == nmsg_res_success);
+	check_return(nmsg_pcap_input_setfilter_raw(pcap, bpf_filter_string) == nmsg_res_success);
+	check_return(nmsg_pcap_input_setfilter(pcap, bpf_filter_string) == nmsg_res_success);
```
Comment (shw700) [inline tests/test-misc.c:499]: Let's keep this second one commented out because it's creating coverage illusion

File: tests/test-kicker.sh [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+[[ "$file_count" == "15" ]] && [[ "$true_file_count" == "16" ]] # 16 to account for json output file.
+check "comparison of kicked json output files"
+
+# Remove kicked files from $outdir.
+while IFS= read -r line; do
+	rm $line
+done < $outfile
+
+# Ensure that a bad kicker script does nothing!  Thank you, quotation marks.
+$nmsgtool_test -ddddd -j $infile -c 1 -k $badkicker > $outfile
```
Comment (shw700) [inline tests/test-kicker.sh:49]: Note that this attempt without quotes produces the following logs:

```
nmsgtool: usage problem, 2 unwanted additional arguments
Usage: nmsgtool
[-b filter]       or --bpf        filter pcap inputs with this bpf
[-B byterate]     or --byterate   ingress byte rate limit for file input
[-c count]        or --count      stop or reopen after count payloads output
```

So in reality nothing is being tested. The test is not even getting past the getopt gateway.

File: tests/test-zmq-endpoints.sh [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+$nmsgtool_test -ddddd --unbuffered -r $infile -S ipc:///tmp/testzmq.sock,pushpull,accept &
+writer_pid=$!
+echo Listening PID: $listener_pid, writing PID: $writer_pid
+
+# Wait for the messages to send/receive.
+sleep 1
+
+# Kill the nmsgtools
+kill $writer_pid >/dev/null 2>&1
+killwriter_pid=$!
```
Comment (shw700) [inline tests/test-zmq-endpoints.sh:39]: More places where there is unnecessary waiting on the kill command

File: tests/test-sample.sh [hunk 11]
```diff
@@ -0,0 +1,35 @@
+#!/bin/bash
+
+
+script_dir=$( cd -- "$( dirname -- "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )
+nmsgtool_test=$script_dir/../src/nmsgtool
+infile=$script_dir/private-tests/202_sample.nmsg
+outfile=/tmp/testsample.json
+entry_count=5
```
Comment (shw700) [inline tests/test-sample.sh:8]: Could use a quick comment explaining how we got this (i.e. 1 out of every 3 x 15 = 5)

File: tests/test-misc.c (hunk 8 above)
Comment (kadealicious) [reply tests/test-misc.c]: Fixed

File: tests/test-kicker.sh (hunk 9 above)
Comment (kadealicious) [reply tests/test-kicker.sh]: Feeling a little silly about this one.  Fixed!

File: tests/test-zmq-endpoints.sh (hunk 10 above)
Comment (kadealicious) [reply tests/test-zmq-endpoints.sh]: Fixed

File: tests/test-socket-frag.sh [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+# Create a listener, wait a second, then create a writer on the same IP/socket.
+$nmsgtool_test -ddddd --unbuffered -m 1 -c $entry_count -o $outfile_frag -l $sockaddr &
+listener_pid=$!
+sleep 1
+$nmsgtool_test -ddddd --unbuffered -m 1 -c $entry_count -p $infile -V base -T dnsqr -s $sockaddr &
+writer_pid=$!
+echo Listening PID: $listener_pid, writing PID: $writer_pid
+
+# Wait for the messages to send/receive.
+sleep 1
```
[the remark this answers is not in the document]
Comment (kadealicious) [reply tests/test-socket-frag.sh]: Sleep is not needed here, fixed

File: tests/test-sample.sh (hunk 11 above)
Comment (kadealicious) [reply tests/test-sample.sh]: Fixed!  Now the sample filter parameter is calculated from entries in the file and desired output entries.

**Review by shw700 — APPROVED** [review]

**Comment by shw700** [comment]
I'm not sure if we should just tag @reedjc in here from the start or wait until it moves past all the devs... I expect he'll be reading through this very closely at some point.

**Review by cmikk — CHANGES_REQUESTED** [review]
Requesting an organization change to move the test files to `generic-tests` or a new `test-data` directory.  Some of the files already present in `generic-tests` may be usable for the basic "plumbing" tests, also.

File: tests/private-tests/202_sample.json
[the remark this answers is not in the document]
Comment (kadealicious) [reply tests/private-tests/202_sample.json:1]: Fixed!

**Review by cmikk — APPROVED** [review]
